src/firmware/app/control/trajectory_planner.c

```diff
@@ -316,22 +358,105 @@ void app_trajectory_planner_generateTimeProfile(
+            path_duration - forwards_trajectory->trajectory_elements[i].time;
+
+        // Reverse the speed profile
+        reverse_speed_profile[(num_segments - 1) - i] =
+            forwards_trajectory->speed_profile[i];
     }
     // Copy reverse element array back
     for (unsigned int i = 0; i < num_segments; i++)
     {
-        forwards[i].position = reverse[i].position;
-        forwards[i].time     = reverse[i].time;
+        forwards_trajectory->trajectory_elements[i].position = reverse[i].position;
+        forwards_trajectory->trajectory_elements[i].time     = reverse[i].time;
+        forwards_trajectory->speed_profile[i]                = reverse_speed_profile[i];
+    }
+}
+
+void app_trajectory_planner_generateVelocityTrajectory(
+    PositionTrajectory_t* position_trajectory, VelocityTrajectory_t* velocity_trajectory)
+{
+    PositionTrajectoryElement_t* position_elements =
+        position_trajectory->trajectory_elements;
+    VelocityTrajectoryElement_t* velocity_elements =
+        velocity_trajectory->trajectory_elements;
+
+    velocity_trajectory->path_parameters = position_trajectory->path_parameters;
+
+    for (unsigned int i = 0; i < position_trajectory->path_parameters.num_segments - 1;
+         i++)
+    {
+        const float delta_x =
+            position_elements[i + 1].position.x - position_elements[i].position.x;
+        const float delta_y =
+            position_elements[i + 1].position.y - position_elements[i].position.y;
+
+        // The unit vector of the direction is 1/magnitide(vector) *vector
+        const float x_velocity_component = (1 / sqrt(pow(delta_x, 2) + pow(delta_y, 2))) *
+                                           delta_x *
+                                           position_trajectory->speed_profile[i];
+        const float y_velocity_component = (1 / sqrt(pow(delta_x, 2) + pow(delta_y, 2))) *
+                                           delta_y *
+                                           position_trajectory->speed_profile[i];
+
+        // Copy data into the velocity element array
+        velocity_elements[i].linear_velocity.x = x_velocity_component;
+        velocity_elements[i].linear_velocity.y = y_velocity_component;
+        velocity_elements[i].angular_velocity  = 0;
+        velocity_elements[i].time              = position_elements[i].time;
     }
+
+    // Assume that the velocity of the final segment is in the direction of the previous
+    const float delta_x =
+        position_elements[position_trajectory->path_parameters.num_segments - 1]
+            .position.x -
+        position_elements[position_trajectory->path_parameters.num_segments - 2]
+            .position.x;
+    const float delta_y =
+        position_elements[position_trajectory->path_parameters.num_segments - 1]
+            .position.y -
+        position_elements[position_trajectory->path_parameters.num_segments - 2]
+            .position.y;
+
+<<<<<<< HEAD
```

Comment:
you got a bad merge here bud :cowboy_hat_face:

Reply:
Fixed 👍